tools/replica/reorder_remd_traj.py

```diff
@@ -0,0 +1,571 @@
+        of.close()
+        buf.close()
+    
+    for i in infobjs: i.close()
+    
+    return
+        
+        
+def get_canonical_logw(enefn, frametuple_dict, temps, nprod, writefreq,
+                       kB = 0.001987):
+    """
+    Gets configurational log-weights (logw) for each frame and at each temp.
+    from the REMD simulation. ONLY WRITTEN FOR THE CANONICAL (NVT) ensemble.
+    
+    This weights can be used to calculate the 
+    ensemble averaged value of any simulation observable X at a given temp. T :
+    <X> (T) = \sum_{k=1, ntemps} \sum_{n=1, nframes} w[idx][k,n] X[k,n]
+    where nframes is the number of frames to use from each *reordered* traj
+    
+    :param enefn: ascii file (readable by numpy.loadtxt) containing an array
+                  u[r,n] of *total* potential energy for the n-th frame for
+                  the r-th replica.
+    
+    :param frametuple_dict: dict containing a tuple (replica #, frame #) 
+                            for each temp.
+    
+    :param temps: array of temps. used in the REMD simulation
+    
+    :param nprod: number of production timesteps. Last (nprod / writefreq) 
+                  frames from the end will be written to disk.
+    
+    :param writefreq: traj dump frequency in LAMMPS
+    
+    :param kB : Boltzmann constant to set the energy scale.
+                Default is in kcal/mol
+    
+    Returns: logw: dict, logw[l][k,n] gives the log weights from the
+                   n-th frame of the k-th temp. *ordered* trajectory
+                   to reweight to the l-th temp.
+    
+    """
+    
+    try:
+        import pymbar
+    except ImportError:
+        print("""
+              Configurational log-weight calculation requires pymbar.
+              Here are some options to install it:
+              conda install -c omnia pymbar
+              pip install pymbar
+          
+              To install the dev. version directly from github, use:
+              pip install pip install git+https://github.com/choderalab/pymbar.git
+              """)
+              
+    u_rn = np.loadtxt(enefn)
+    ntemps = u_rn.shape[0] # number of temps.
+    nframes = int(nprod / writefreq) # number of frames at each temp.
+    
+    # reorder the temps
+    u_kn = np.zeros([ntemps, nframes], float)
+    for k in range(ntemps):
+        frame_tuple = frametuple_dict[k][-nframes:]
+        for i, (rep, frame) in enumerate(frame_tuple):
+            u_kn[k, i] = u_rn[rep, frame]
+    
+    # prep input for pymbar    
+    #1) array of frames at each temp.
+    nframes_k = nframes * np.ones(ntemps, np.uint8)
+    
+    #2) inverse temps. for chosen energy scale
+    beta_k = 1.0 / (kB * temps)
+    
+    #3) get reduced energies (*ONLY FOR THE CANONICAL ENSEMBLE*)
+    u_kln = np.zeros([ntemps, ntemps, nframes], float)
+    for k in range(ntemps):
+        for l in range(ntemps):
+            u_kln[ k, l, 0:nframes_k[k] ] = beta_k[l] * u_kn[k, 0:nframes_k[k]]
+    
```

Comment:
It would be nice to avoid the double loop and speed up the calculations. Maybe this would be faster
for k in range(ntemps):
   u_kln[ k, :, 0:nframes_k[k] ] = beta_k[:] * u_kn[k, 0:nframes_k[k]]

Reply:
The vectorization you suggested leads to a broadcasting error. I used numpy's outer product instead.
for k in range(ntemps):
u_kln[k]  = np.outer(beta_k, u_kn[k])

and that works fine

Reply:
Do you see any speedup ?

Reply:
Not really for the small system I am running tests on. Needs to be tested on something large.